File: ra-idp-go/internal/oauth2/domain/refresh_token.go
// refresh token の domain。ADR-004 のローテーション + ファミリー失効の基盤。
package domain

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"time"

	"ra-idp-go/internal/spec"
)

const (
	refreshTokenBytes       = 48
	refreshTokenTTL         = 14 * 24 * time.Hour
	refreshTokenAbsoluteTTL = 30 * 24 * time.Hour
)

type GeneratedRefreshToken struct {
	Token  string
	Record *spec.RefreshTokenRecord
}

func HashRefreshToken(token string) string {
	sum := sha256.Sum256([]byte(token))
	return hex.EncodeToString(sum[:])
}

func GenerateInitialRefreshToken(clientID, sub string, scopes []string, sc *spec.SenderConstraint, now time.Time) (*GeneratedRefreshToken, error) {
	if now.IsZero() {
		now = time.Now().UTC()
	}
	b := make([]byte, refreshTokenBytes)
	if _, err := rand.Read(b); err != nil {
		return nil, err
	}
	token := base64.RawURLEncoding.EncodeToString(b)
	id, err := spec.NewUUIDv4()
	if err != nil {
		return nil, err
	}
	familyID, err := spec.NewUUIDv4()
	if err != nil {
		return nil, err
	}
	rec := &spec.RefreshTokenRecord{
		ID:                id,
		Hash:              HashRefreshToken(token),
		FamilyID:          familyID,
		ClientID:          clientID,
		Sub:               sub,
		Scopes:            scopes,
		IssuedAt:          now,
		ExpiresAt:         now.Add(refreshTokenTTL),
		AbsoluteExpiresAt: now.Add(refreshTokenAbsoluteTTL),
		SenderConstraint:  sc,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	return &GeneratedRefreshToken{Token: token, Record: rec}, nil
}

func RotateRefreshToken(parent *spec.RefreshTokenRecord, now time.Time) (*GeneratedRefreshToken, error) {
	if now.IsZero() {
		now = time.Now().UTC()
	}
	b := make([]byte, refreshTokenBytes)
	if _, err := rand.Read(b); err != nil {
		return nil, err
	}
	token := base64.RawURLEncoding.EncodeToString(b)
	id, err := spec.NewUUIDv4()
	if err != nil {
		return nil, err
	}
	// expires_at は短い方を採用 (absolute_expires_at で打ち切り)。
	expires := now.Add(refreshTokenTTL)
	if expires.After(parent.AbsoluteExpiresAt) {
		expires = parent.AbsoluteExpiresAt
	}
	parentID := parent.ID
	rec := &spec.RefreshTokenRecord{
		ID:                id,
		Hash:              HashRefreshToken(token),
		FamilyID:          parent.FamilyID,
		ParentID:          &parentID,
		ClientID:          parent.ClientID,
		Sub:               parent.Sub,
		Scopes:            parent.Scopes,
		IssuedAt:          now,
		ExpiresAt:         expires,
		AbsoluteExpiresAt: parent.AbsoluteExpiresAt,
		SenderConstraint:  parent.SenderConstraint,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	return &GeneratedRefreshToken{Token: token, Record: rec}, nil
}

func IsRefreshTokenReplay(rec *spec.RefreshTokenRecord) bool {
	return rec.Rotated || rec.Revoked
}

func IsRefreshTokenAbsoluteExpired(rec *spec.RefreshTokenRecord, now time.Time) bool {
	if now.IsZero() {
		now = time.Now()
	}
	return !now.Before(rec.AbsoluteExpiresAt)
}

File: ra-idp/src/oauth2/domain/refresh-token.ts
/**
 * Layer 3 — Application Logic
 *
 * リフレッシュトークンのドメインモデル。
 * ADR-004 に従いローテーション + ファミリー失効を実装する基盤を提供する。
 *
 * 仕様核 `spec/scl.yaml` の models.RefreshTokenRecord（派生: `gen/RefreshTokenRecord.json`）の制約に従う。
 * トークン文字列自体は不透明 (opaque) で、ストアレコードに SHA-256 ハッシュとして保存する。
 */

import { randomBytes, createHash, randomUUID } from 'crypto'
import { RefreshTokenRecordSchema, type RefreshTokenRecord } from '../../spec-bindings/schemas'

export type { RefreshTokenRecord }

const TOKEN_BYTES = 48 // 384 ビット
const DEFAULT_TTL_SECONDS = 60 * 60 * 24 * 14 // 14 日
const ABSOLUTE_TTL_SECONDS = 60 * 60 * 24 * 30 // 30 日（ADR-004）

export interface GeneratedRefreshToken {
  /** クライアントに返却するトークン文字列（不透明）。 */
  token: string
  /** ストアに保存するレコード。 */
  record: RefreshTokenRecord
}

export function hashToken(token: string): string {
  return createHash('sha256').update(token).digest('hex')
}

export function generateInitial(input: {
  tenant_id: string
  client_id: string
  sub: string
  scopes: string[]
  sender_constraint?: RefreshTokenRecord['sender_constraint']
  now?: Date
}): GeneratedRefreshToken {
  const now = input.now ?? new Date()
  const token = randomBytes(TOKEN_BYTES).toString('base64url')
  const familyId = randomUUID()
  const record = RefreshTokenRecordSchema.parse({
    id: randomUUID(),
    tenant_id: input.tenant_id,
    hash: hashToken(token),
    family_id: familyId,
    client_id: input.client_id,
    sub: input.sub,
    scopes: input.scopes,
    issued_at: now.toISOString(),
    expires_at: new Date(now.getTime() + DEFAULT_TTL_SECONDS * 1000).toISOString(),
    absolute_expires_at: new Date(now.getTime() + ABSOLUTE_TTL_SECONDS * 1000).toISOString(),
    revoked: false,
    rotated: false,
    sender_constraint: input.sender_constraint ?? null,
  })
  return { token, record }
}

export function rotate(parent: RefreshTokenRecord, now: Date = new Date()): GeneratedRefreshToken {
  const token = randomBytes(TOKEN_BYTES).toString('base64url')
  const record = RefreshTokenRecordSchema.parse({
    id: randomUUID(),
    tenant_id: parent.tenant_id,
    hash: hashToken(token),
    family_id: parent.family_id,
    parent_id: parent.id,
    client_id: parent.client_id,
    sub: parent.sub,
    scopes: parent.scopes,
    issued_at: now.toISOString(),
    // expires_at は再延長、ただし absolute_expires_at は親から継承
    expires_at: new Date(
      Math.min(now.getTime() + DEFAULT_TTL_SECONDS * 1000, Date.parse(parent.absolute_expires_at)),
    ).toISOString(),
    absolute_expires_at: parent.absolute_expires_at,
    revoked: false,
    rotated: false,
    sender_constraint: parent.sender_constraint ?? null,
  })
  return { token, record }
}

export function isReplay(record: RefreshTokenRecord): boolean {
  return record.rotated || record.revoked
}

export function isAbsoluteExpired(record: RefreshTokenRecord, now: Date = new Date()): boolean {
  return now.getTime() >= Date.parse(record.absolute_expires_at)
}
